### ISSUE TITLE
Is there any default implementation of shuffling?

### ISSUE THREAD

**Issue by Jokeren** [NervanaSystems/neon#49]
I realize that shuffling could be an essential part in NN. However, some people indicate that "If you are using a not-recurrent NN like a traditional MLP you don't NEED to shuffle dataset especially if you're using a batch learning algorithm". 

I wonder whether there's a shuffling step before each epoch in **neon**? If not, how can I implement it?

Related articles:

http://stats.stackexchange.com/questions/40638/predicting-time-series-with-nns-should-the-data-set-be-shuffled

http://stats.stackexchange.com/questions/90874/how-can-stochastic-gradient-descent-avoid-the-problem-of-a-local-minimum

**Comment by scttl** [comment]
Hi,

By default there is no such shuffling before each epoch in neon.  Most of the models are setup to run mini-batched gradient descent, where the `batch_size` model parameter defines how many examples are in each batch.

If you wanted to shuffle mini-batch order within an epoch that should be pretty straight forward and I'd recommend having a look in the `DataLayer` class code.  I'd suggest creating a subclass and overriding the `has_more_data` and `fprop` functions.  If however you wanted to shuffle individual examples this will be more complex and result in a lot of overhead (as data is stored on device in mini-batches and accessed as such).

**Comment by apark263** [comment]
We have capability for shuffling within a macrobatch for imageset based
datasets. It hasn't been pushed to external master yet, but it could be a
good compromise between full set shuffling and just presenting the same
minibatches on each epoch.

**Comment by Jokeren** [comment]
Thank you!

@apark263, after reading codes these days, I notice what you mentioned about the `Imagest`.

I am currently working on modifying `Imageset` for my own tasks. `Imageset` is a useful tool to read large set of images from the disk, whereas there are some additional features we do not need such as cropping images. Furthermore, it can not be used for testing data and pictures other than `jpeg`.

I guess the `imageset` is originally designed for contest dataset such as imagenet?

Maybe there could be a another class, which is the subclass of the `Dataset` and the super class of `ImageSet`, only provides the feature of reading huge amount of images. Or we should modify the `batch_writer`?

**Comment by apark263** [comment]
Imageset was originally meant to support training and testing of ImageNet
1k models, however we have tried to make it general enough to handle the
case that you are describing.

Most of the modifications you are looking for can be handled by supplying
the appropriate options to batchwriter, which preprocesses the images into
macrobatches (it also converts non-jpg images to jpegs for decoding by
Imageset). The types of formats that batchwriter can support are whatever
image formats that can be read by PIL.

Regarding not wanting to use random cropping or flipping: If you just
preprocess the images to the final size that you want to test, then set
cropped_image_size and output_image_size to that fixed image size, then no
cropping will occur. If you also specify dotransforms=False, no flips will
occur.
